### ISSUE TITLE
Making sense of the depth texture data

### ISSUE THREAD

**Issue by SergioRAgostinho** [cprogrammer1994/ModernGL#199]
### Issue / Question / Feature request

~~Question~~ Issue

### Version info

```
$ python3 -m moderngl
moderngl 5.2.1
--------------
vendor: NVIDIA Corporation
renderer: GeForce GTX 960/PCIe/SSE2
version: 4.5.0 NVIDIA 384.130
python: 3.5.2 (default, Nov 23 2017, 16:37:01) 
[GCC 5.4.0 20160609]
platform: linux
code: 450
```

### Description

As indicated, what I have is a question that with some luck and your expertise might be quickly solved. I'm rendering a scene into a framebuffer object. Everything is being rendered correctly and I'm able to read out the framebuffer into a numpy array and correctly get all the color values for each pixel. So far so good.

My next step was to retrieve the depth values of this scene. For that I used a texture as a depth attachment, because the renderbuffer doesn't have a read method. The "metadata" seems to be correct
```
(Pdb) dbo.size
(640, 480)
(Pdb) dbo.components
1
(Pdb) dbo.dtype
'f4'
```

Which made me assume that the depth buffer keeps a 4 byte floating point per pixel. Also cool.
Then came the type to read values into a numpy array to send them out. My assumption was that the depth buffer should hold one 32bit floating point per pixel which should range between -1 and 1, since the projection to Normalized Device Cordinates (NDC) enforces this. To read the data into the numpy an array I was using
```python
dmp = np.frombuffer(dbo.read(), dtype=np.dtype('f4')).reshape((480, 640))
```
where `dbo` is the depth_texture. Upon probing the numpy array image for its maximum and minimum then I got some weird results
```
(Pdb) dmp = np.frombuffer(dbo.read(), dtype=np.dtype('f4')).reshape((480, 640))
(Pdb) dmp
array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)
(Pdb) np.max(dmp)
nan
(Pdb) np.min(dmp)
/home/sergio/.local/lib/python3.5/site-packages/numpy/core/_methods.py:29: RuntimeWarning: invalid value encountered in reduce
  return umr_minimum(a, axis, None, out, keepdims)
nan
```
There's no way the depth buffer should hold nans, since it was cleared with depth set to 1 (far). Then something raised my attention which is the `alignment` setting that Texture objects. This can be set on creation and on the read method. I tried also reading with alignment=4 but didn't yield good results

```
(Pdb) dmp = np.frombuffer(dbo.read(alignment=4), dtype=np.dtype('f4')).reshape((480, 640))
(Pdb) dmp
array([[-1.7014118e+38, -1.7014118e+38, -1.7014118e+38, ...,
        -1.7014118e+38, -1.7014118e+38, -1.7014118e+38],
       [-1.7014118e+38, -1.7014118e+38, -1.7014118e+38, ...,
        -1.7014118e+38, -1.7014118e+38, -1.7014118e+38],
       [-1.7014118e+38, -1.7014118e+38, -1.7014118e+38, ...,
        -1.7014118e+38, -1.7014118e+38, -1.7014118e+38],
       ...,
       [-1.7014118e+38, -1.7014118e+38, -1.7014118e+38, ...,
        -1.7014118e+38, -1.7014118e+38, -1.7014118e+38],
       [-1.7014118e+38, -1.7014118e+38, -1.7014118e+38, ...,
        -1.7014118e+38, -1.7014118e+38, -1.7014118e+38],
       [-1.7014118e+38, -1.7014118e+38, -1.7014118e+38, ...,
         6.5957717e-41,  0.0000000e+00,  0.0000000e+00]], dtype=float32)
(Pdb) np.max(dmp)
1.88179e-40
(Pdb) np.min(dmp)
-2.3686692e+38
```
I tried different combinations of the alignment setting during the creation and reading of the depth texture but no success. At this point it became clear that I have little understanding of the role of the alignment keyword and I decided to raise the question here hoping for good ideas on how to proceed.

I mention again the ultimate goal is to retrieve the depth value from the buffer, which I assumed that is should be between -1 and 1.

Thanks

**Comment by einarf** [comment]
I've solved this in the past by rendering the depth buffer to a float texture with one component. This was more a convenience thing as I wanted to linearlize the values and did not want to care what format the depth buffer was exactly.

**Comment by SergioRAgostinho** [comment]
'sup Einarf :) So far my workaround has been to write a(n OpenGL) program to render the depth into a greyscale image, which is perfectly fine. I just thought since a depth map is being generated anyway in the other programs I'm using, I could avoid having the extra step.

Regarding the alignment: I'm not passing any data upon creation of the depth_texture (I don't think it's required) so I assume the `GL_PACK_ALIGNMENT` should have no effect here; upon reading data from the depth_texture I have no alignment guaranties regarding the destination memory of my numpy array. I've set the alignment to 1 in both depth_texture creation and read operations for the time being.

On other topics, the line immediately below the ones you posted got my attention
https://github.com/cprogrammer1994/ModernGL/blob/9a5d09a2ebada8ff980be8137184ff2b66a6fa7f/src/Texture.cpp#L246

which tells me the depth buffer is actually set to 24bits although the we're defined the `pixel_type` as `GL_FLOAT`. Floats are usually 32bits in most desktops, so I'm wondering exactly how is this depth being stored.

Just for safety I'm putting more code detailing the creation and read of the depth_texture
```python
size = (640, 480)
cbo = self._ctx.renderbuffer(size)
dbo = self._ctx.depth_texture(size, alignment=1)
fbo = self._ctx.framebuffer(color_attachments=(cbo), depth_attachment=dbo)
fbo.use()
fbo.clear(red=0.0, green=0.0, blue=0.0, alpha=1.0, depth=1.0)

# [code for setting uniforms and vertex array rendering]

# time to get things from the fbo
dmp = np.frombuffer(dbo.read(alignment=1), dtype=np.dtype('f4')).reshape((480, 640))
```

**Comment by SergioRAgostinho** [comment]
> Do you have DEPTH_TEST enabled?

Yes and it's working perfectly.

> HAve you tried to write a value explicitly to gl_FragDepth? Can you read it back?

No. I'll give it a try in a self contained example and report results.

**Comment by SergioRAgostinho** [comment]
Here's a self contained example for us to work on. It exploits the same concepts I described before. I'm not setting the depth manually, but something is clearly being rendered as you can tell from the printed output.

```python
import numpy as np
import moderngl as mgl

ctx = mgl.create_standalone_context()
ctx.enable(mgl.DEPTH_TEST)
prog = ctx.program(
    vertex_shader='''
        #version 330

        in vec3 in_vert;

        void main() {
            gl_Position = vec4(in_vert, 1.0);
        }
    ''',
    fragment_shader='''
        #version 330

        out vec3 f_color;

        void main() {
            f_color = vec3(1.0);
        }
    ''',
)

vertices = np.array([
    [-1, 1, 0.],
    [-1, -1, 0.],
    [1, -1, 0.],
    ])
vbo = ctx.buffer(vertices.astype('f4').tobytes())

size = (4, 4)
cbo = ctx.renderbuffer(size)
dbo = ctx.depth_texture(size, alignment=1)

fbo = ctx.framebuffer(color_attachments=(cbo), depth_attachment=dbo)
fbo.use()
fbo.clear(depth=1.0)

depth = np.frombuffer(dbo.read(alignment=1), dtype=np.dtype('f4')).reshape(size[::-1])
print(depth) # should all be 1

# Now we render a triangle in there
vao = ctx.simple_vertex_array(prog, vbo, 'in_vert')
vao.render()

# It should have some different values in the middle
depth = np.frombuffer(dbo.read(alignment=1), dtype=np.dtype('f4')).reshape(size[::-1])
print(depth) # should all be 1

# lets just check the value of the fbo for safety
data = np.frombuffer(fbo.read(), dtype=np.dtype('u1')).reshape((size[1], size[0], 3))
print(data[:,:,0])
```

Yields the following
```
$ ./depth_example.py 
[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]
[[9.0074441e+15 9.1835510e-39 0.0000000e+00 0.0000000e+00]
 [6.7421017e+22 2.6393955e-06 2.0617684e-19 2.0706459e-19]
 [7.3971504e+20 1.7595148e+22 7.1220781e+28 1.8264451e+34]
 [5.8371669e-10 7.3988117e+31 1.8370126e+25 2.3340747e-09]]
[[255 255 255   0]
 [255 255   0   0]
 [255   0   0   0]
 [  0   0   0   0]]
```

The first output should be all 1s, the second 1s with something between 0 and 1 where the triangle is. The third output is just there to prove the triangle is being rendered.

**Comment by cprogrammer1994** [comment]
Hi! Reading a depth texture cause an OpenGL error. I will have to investigate this. I will add a new method, or extend the read method of the Framebuffer. Depth values should be read through a framebuffer as I see now. I was not able to read depth values correctly wih `glGetTexImage`

**Comment by cprogrammer1994** [comment]
Hi! This is a bit [hacky](https://github.com/cprogrammer1994/ModernGL/commit/5434172da890a95566c3720a76ec007f7faded9a) but solves the problem. I will prepare a patch that will add support for reading the depth values. One must use a framebuffer to read depth values from a texture (an error will be raised when read directly).

**Comment by cprogrammer1994** [comment]
This bug should be fixed in the next release.

**Comment by cprogrammer1994** [comment]
There is a fix in #219

**Comment by cprogrammer1994** [comment]
merged #219. it also works with `Texture.read` and `Texture.read_into`
`attachment=1` refers to the depth attachment when reading a framebuffer.
